### PR TITLE
[Model] Add GraphCare — KG-enhanced EHR predictions (ICLR 2024)

### PR DESCRIPTION
**Pull request by joshuasteier**
<html>
<body>
<html><head></head><body><h1>[Model] Add GraphCare — KG-enhanced EHR predictions (ICLR 2024)</h1>
<h2>Contributor</h2>
<ul>
<li><strong>Name:</strong> Josh Steier</li>
</ul>
<h2>Contribution Type</h2>
<p>New Model</p>
<h2>Description</h2>
<p>Implements <strong>GraphCare</strong> from <em>"GraphCare: Enhancing Healthcare Predictions with Personalized Knowledge Graphs"</em> (Jiang et al., ICLR 2024).</p>
<p>GraphCare constructs personalized patient-level knowledge graphs from EHR codes (conditions, procedures, drugs) and applies a GNN with bi-attention pooling for downstream prediction tasks. This is a faithful port of the <a href="https://github.com/pat-jj/GraphCare">original implementation</a> adapted for PyHealth conventions.</p>
<p><strong>Key features:</strong></p>
<ul>
<li>Three GNN backbones: BAT (Bi-Attention GNN, from the paper), GAT, GIN</li>
<li>Three patient representation modes: <code>joint</code>, <code>graph</code>, <code>node</code></li>
<li>Visit-level (alpha) and node-level (beta) attention with temporal decay</li>
<li>Optional edge attention and edge dropout</li>
<li>Attention weight extraction for interpretability (<code>store_attn=True</code>)</li>
<li>Support for pre-trained or learned node/relation embeddings</li>
</ul>
<p><strong>Supported tasks:</strong> mortality, readmission, drug recommendation, length-of-stay</p>
<p><strong>Paper:</strong> https://openreview.net/forum?id=tVTN7Zs0ml<br>
<strong>Original repo:</strong> https://github.com/pat-jj/GraphCare</p>
<h2>⚠️ Points for Maintainer Discussion</h2>
<h3>1. <code>torch-geometric</code> dependency</h3>
<p>This is the first PyHealth model that requires <code>torch-geometric</code>. It is handled as a <strong>lazy optional dependency</strong> — imported inside a try/except with a clear error message if missing. No changes to <code>setup.py</code> or <code>requirements.txt</code> are needed, but I wanted to flag this for your review. Happy to adjust the approach.</p>
<h3>2. Does not inherit from <code>BaseModel</code></h3>
<p>GraphCare operates on <code>torch_geometric.data.Data</code> objects (per-patient subgraphs) rather than PyHealth's standard <code>SampleDataset</code> collation. This means it cannot use PyHealth's <code>Trainer</code> directly. The model is a standalone <code>nn.Module</code> with its own data pipeline utilities (<code>graphcare_utils.py</code>). This mirrors how the original paper's pipeline works. Open to suggestions on integration.</p>
<h3>3. Synthetic data in tests and tutorial</h3>
<p>The unit tests and tutorial notebook use <strong>synthetic patient graphs</strong> rather than real MIMIC-III data. This is intentional — the real KG artifacts require MIMIC access plus running the full LLM-prompted KG construction pipeline from the original repo. The synthetic data validates architecture correctness (shapes, gradients, no NaN across all 9 GNN×mode combinations). Section 9 of the tutorial shows the real-data workflow using <code>graphcare_utils</code>.</p>
<h2>Files to Review</h2>

File | Path | Description
-- | -- | --
Model | pyhealth/models/graphcare.py | GraphCare + BiAttentionGNNConv (507 lines)
Data utils | pyhealth/models/graphcare_utils.py | KG loading, subgraph extraction, DataLoader helpers
Unit tests | tests/core/test_graphcare.py | 28 tests covering all GNN/mode combos, attention flags, edge dropout, embeddings, numerical sanity
Tutorial | examples/graphcare_tutorial.ipynb | End-to-end: synthetic data → train → evaluate → attention visualization
Docs | docs/api/models/pyhealth.models.GraphCare.rst | Sphinx autodoc page
Docs index | docs/api/models.rst | Added toctree entry (between GAMENet and GRASP)
Init | pyhealth/models/__init__.py | Added import line

<h2>How to Test</h2>
<pre><code class="language-bash"># Run unit tests (requires torch-geometric)
pip install torch-geometric
python -m pytest tests/test_graphcare.py -v

# Run tutorial notebook
jupyter notebook examples/graphcare_tutorial.ipynb
</code></pre>
<h2>Checklist</h2>
<ul>
<li>[x] PEP8 compliant, 88-char line limit</li>
<li>[x] Google-style docstrings with type hints</li>
<li>[x] Unit tests with dummy inputs and expected outputs (28 tests)</li>
<li>[x] Example notebook with end-to-end usage</li>
<li>[x] Sphinx documentation page</li>
<li>[x] Rebased with <code>sunlabuiuc/PyHealth</code> main</li>
<li>[x] "Allow edits by maintainers" enabled</li>
</ul></body></html>
</body>

### CODE REVIEW

File: docs/api/models/pyhealth.models.GraphCare.rst
Comment (jhnwu3) [inline docs/api/models/pyhealth.models.GraphCare.rst:1]: I'm not sure why this only has "white space only changes" here. Maybe a typo?

**Comment by joshuasteier** [comment]
Hi, @jhnwu3:

Definitely, I'd love to build a MIMIC-native version of this once I have data/compute access. I think the natural path is something like:

1. This PR as-is (standalone with its own PyG pipeline, works for people who already have KG artifacts from the original repo)
2. A `PyHealthGraphDataset` adapter that takes a standard `SampleDataset` and converts patient visits into PyG `Data` objects — subgraph extraction, code-to-KG mapping, DataLoader-ready batches. This is probably currently the most reasonable next step.
3. Longer term, native PyG support in PyHealth's core so any EHR dataset can be projected onto a KG and consumed by graph models (GraphCare, G-BERT, KAME, etc.)

I'd want to take on step 2. The main design question is where KG construction fits — pre-computed artifacts vs. on-the-fly from code mappings — and how that interacts with `SampleDataset`. Let me know if you'd prefer to hash that out async or on a call.

---

# Reply 2: Whitespace-only changes

Good catch, will fix.

Thank you!

**Comment by jhnwu3** [comment]
re: I think the better approach is probably building a pyhealth.graph module and adding pyhealth graph processors to pyhealth.processors. Let me see if I can't find Patrick's old branch. 

https://github.com/sunlabuiuc/PyHealth/tree/kg_embedding/pyhealth/datasets 
https://github.com/sunlabuiuc/PyHealth/blob/kg_embedding/pyhealth/datasets/base_kg_dataset.py


Patrick had an old base_kg_dataset that's basically this idea I believe before we had all of these backend changes. Let's talk more later this week since I don't exactly have a perfect idea of how best to define these graph structures or store them within pyhealth. (Maybe we just assume the user has to provide it in some format).

**Comment by joshuasteier** [comment]
Superseded by the new implementation that uses the pyhealth.graph infrastructure from #853. Resubmitting as a clean PR that inherits BaseModel and consumes KnowledgeGraph + GraphProcessor natively.
